src/io.jl

```diff
@@ -199,10 +199,7 @@ function _load(png_ptr, info_ptr; gamma::Union{Nothing,Float64}=nothing, expand_
         PNG_HEADER_VERSION_STRING
     )
 
-    png_read_image(png_ptr, map(pointer, eachcol(buffer)))
-    png_read_end(png_ptr, info_ptr)
-    png_destroy_read_struct(Ref{Ptr{Cvoid}}(png_ptr), Ref{Ptr{Cvoid}}(info_ptr), C_NULL)
-    buffer = permutedims(buffer, (2, 1))
+    buffer = Base.invokelatest(_load!, buffer, png_ptr, info_ptr)
```

Comment:
I've not quite got the hang of when `invokelatest` is needed. Why would it be needed here?

Reply:
So, this uses "the kernel trick" to allow it to be optimized on the actual element type of the matrix. However, without `invokelatest`, inference will still infer `_load!` with an abstract element type, and unfortunately I seem to remember that wasn't cacheable in the `*.ji` file (although I could have just forgotten to add it to the precompiles, so take that with a giant grain of salt).

Calling it by `invokelatest` ensures that you only ever infer it for concrete type. Demo:
```julia
julia> include("runtests.jl")
< suppressed output>

julia> using MethodAnalysis

julia> methodinstances(PNGFiles._load!)
9-element Vector{Core.MethodInstance}:
 MethodInstance for _load!(::Matrix{UInt8}, ::Ptr{Nothing}, ::Ptr{Nothing})
 MethodInstance for _load!(::Matrix{Gray{N0f8}}, ::Ptr{Nothing}, ::Ptr{Nothing})
 MethodInstance for _load!(::Matrix{Gray{N0f16}}, ::Ptr{Nothing}, ::Ptr{Nothing})
 MethodInstance for _load!(::Matrix{GrayA{N0f8}}, ::Ptr{Nothing}, ::Ptr{Nothing})
 MethodInstance for _load!(::Matrix{GrayA{N0f16}}, ::Ptr{Nothing}, ::Ptr{Nothing})
 MethodInstance for _load!(::Matrix{RGB{N0f8}}, ::Ptr{Nothing}, ::Ptr{Nothing})
 MethodInstance for _load!(::Matrix{RGB{N0f16}}, ::Ptr{Nothing}, ::Ptr{Nothing})
 MethodInstance for _load!(::Matrix{RGBA{N0f8}}, ::Ptr{Nothing}, ::Ptr{Nothing})
 MethodInstance for _load!(::Matrix{RGBA{N0f16}}, ::Ptr{Nothing}, ::Ptr{Nothing})
```

You'd have at least one abstract signature in that list without the `invokelatest`. Since there doesn't seem to be any real point in having it, `invokelatest` didn't seem to have any downsides.

Reply:
Awesome. Thanks for the explanation and for making the tools so easy to use. I'll try to keep an eye out for this kind of thing

Reply:
To be fair, it probably is a minor point (in this case just one extra MethodInstance), so I am unsure myself how often I'll be doing that in the future. I suspect we're entering an era where we'll be collectively feeling our way forward on such issues.